### PR TITLE
05_28_2022/gcbui/offscreen_locomotion

### PR DESCRIPTION
**Pull request by gcbui**
offscreen objects

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1,7 +1,32 @@
+import asyncio
+import sys
+import msvcrt
+
 height = 40 #40 characters
 width = 172 #171 characters
 skip_lines = 2
 height = height - skip_lines
+world_x = -20
+world_y = 20
```
Comment (BolajiMakinde) [inline main.py]: all caps for global keywords

File: main.py [hunk 2]
```diff
@@ -1,7 +1,32 @@
...
+    elif move == 's':
+        world_y-=1
+    elif move == 'a':
+        world_x-=1
+    elif move == 'q':
+        quit()
+
+
+
+
```
Comment (BolajiMakinde) [inline main.py]: remove extra spaces. let's use only 1 space between ALL our functions

File: main.py [hunk 3]
```diff
@@ -1,7 +1,32 @@
...
+import msvcrt
+
 height = 40 #40 characters
 width = 172 #171 characters
 skip_lines = 2
 height = height - skip_lines
+world_x = -20
+world_y = 20
+
+def get_input():
```
Comment (BolajiMakinde) [inline main.py]: move function under all the classes.

**Review by BolajiMakinde — CHANGES_REQUESTED** [review]

File: main.py [hunk 4]
```diff
@@ -1,7 +1,32 @@
+import asyncio
+import sys
+import msvcrt
+
 height = 40 #40 characters
 width = 172 #171 characters
```
Comment (BolajiMakinde) [inline main.py]: check to see if we can use this for height and width: https://www.geeksforgeeks.org/python-os-get_terminal_size-method/#:~:text=os.get_terminal_size%20%28%29%20method%20in%20Python%20is%20used%20to,represents%20height%20of%20the%20terminal%20window%20in%20characters.

File: main.py [hunk 5]
```diff
@@ -24,31 +49,52 @@ def init_background(self, background_symbol, width, height):
...
         draw(width/2,height/2,symbol='c')
         draw(1,1,symbol='L')
         draw(width,height,symbol='R')
         draw(width,1,symbol='d')
         draw(1,height,symbol='u')
-        draw(width/2,height/2,symbols=character_hash_map)
+        draw(width/2,height/2,symbols=character_hash_map,use_screen_space=True)
+        draw(1,height/2,symbols=object_hash_map)
         render_frame()
+
```
Comment (BolajiMakinde) [inline main.py:103]: add a function called 

> update_frame_dimensions
that uses the terminal height and width and is called in every for loop

File: main.py [hunk 6]
```diff
@@ -24,31 +49,52 @@ def init_background(self, background_symbol, width, height):
...
     print(current_frame.banner+current_frame.image)
 
 def game():
+    global world_x
+    global world_y
     character_hash_map = {tuple([2,1]):'|', tuple([8,1]):'|', tuple([3,2]):'_',tuple([4,2]):'_',tuple([5,2]):'|',tuple([6,2]):'_',tuple([7,2]):'_', tuple([1,3]):'|',tuple([5,3]):'|',tuple([9,3]):'|',tuple([2,4]):'_', tuple([3,4]): '_',tuple([4,4]):'_',tuple([5,4]):'|',tuple([6,4]):'_',tuple([7,4]):'_',tuple([8,4]):'_', tuple([5,5]):'>'                        
 ,tuple([4,6]):'0',tuple([6,6]):'0',tuple([4,7]):'~',tuple([6,7]):'~'}
-
+    object_hash_map = {tuple([1,1]):'_', tuple([2,1]):'_', tuple([3,1]):'_',tuple([1,2]):'|',tuple([3,2]):'|',tuple([1,3]):'_',tuple([2,3]):'_', tuple([3,3]):'_'}
     for num in range(10000):
```
Comment (BolajiMakinde) [inline main.py]: make this a while true statement now

File: main.py [hunk 7]
```diff
@@ -1,7 +1,32 @@
+import asyncio
+import sys
+import msvcrt
+
 height = 40 #40 characters
```
Comment (BolajiMakinde) [inline main.py]: all variables outside of a function are global so let's also make width, height, etc all caps

File: main.py [hunk 8]
```diff
@@ -24,31 +49,52 @@ def init_background(self, background_symbol, width, height):
...
+        if screen_x <1 or screen_y <1 or screen_x > width or screen_y > height:
+            return
+        i = int((height-screen_y) * (width) + (screen_x-1))
         current_frame.image = current_frame.image[:i] + symbol + current_frame.image[i + 1:]
     else:
         for symbol in symbols:
-            i = int((height-y-symbol[1]+1) * (width) + (x+symbol[0]-2))
+            screen_x = x+symbol[0]-1
+            screen_y = y+symbol[1]-1
+            if use_screen_space == False:
```
Comment (BolajiMakinde) [inline main.py]: change to `if not use_screen_space:`

File: main.py [hunk 9]
```diff
@@ -24,31 +49,52 @@ def init_background(self, background_symbol, width, height):
...
 
-def draw(x,y,symbol=None,symbols=None):
+def draw(x,y,symbol=None,symbols=None,use_screen_space=False):
+    global world_x
+    global world_y
     if symbols == None:
-        i = int((height-y) * (width) + (x-1))
+        screen_x = x 
+        screen_y = y
+        if use_screen_space == False:
```
Comment (BolajiMakinde) [inline main.py]: change to `if not use_screen_space:`

File: main.py [hunk 10]
```diff
@@ -1,7 +1,32 @@
+import asyncio
```
Comment (BolajiMakinde) [inline main.py]: try your best to resolve as many of the errors in PyLint. under the conversation tab at the place where you merge there should be a link to view the PyLint compiler errors and warnings. Try to resolve as many of them as possible.

File: main.py [hunk 11]
```diff
@@ -24,31 +49,52 @@ def init_background(self, background_symbol, width, height):
       
 current_frame = Frame()
 
-def draw(x,y,symbol=None,symbols=None):
+def draw(x,y,symbol=None,symbols=None,use_screen_space=False):
+    global world_x
+    global world_y
     if symbols == None:
```
Comment (BolajiMakinde) [inline main.py]: change to `if symbols is None:`

File: main.py [hunk 12]
```diff
@@ -24,31 +49,52 @@ def init_background(self, background_symbol, width, height):
...
         draw(width,height,symbol='R')
         draw(width,1,symbol='d')
         draw(1,height,symbol='u')
-        draw(width/2,height/2,symbols=character_hash_map)
+        draw(width/2,height/2,symbols=character_hash_map,use_screen_space=True)
+        draw(1,height/2,symbols=object_hash_map)
         render_frame()
+
        
 game()
```
Comment (BolajiMakinde) [inline main.py:105]: remove all the extra spaces

File: main.py [hunk 13]
```diff
@@ -1,7 +1,32 @@
+import asyncio
+import sys
+import msvcrt
```
Comment (BolajiMakinde) [inline main.py:1]: we do not need the `asyncio` and `sys` libraries anymore.

**Review by BolajiMakinde — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (gcbui) [reply main.py]: Replaced with all caps

File: main.py (hunk 2 above)
Comment (gcbui) [reply main.py]: Removed spaces

File: main.py (hunk 3 above)
Comment (gcbui) [reply main.py]: Moved functions under

File: main.py (hunk 6 above)
Comment (gcbui) [reply main.py]: Changed to while true

File: main.py (hunk 7 above)
Comment (gcbui) [reply main.py]: Changed global variables to all caps

File: main.py (hunk 8 above)
Comment (gcbui) [reply main.py]: Changed

File: main.py (hunk 9 above)
Comment (gcbui) [reply main.py]: Changed

File: main.py (hunk 11 above)
Comment (gcbui) [reply main.py]: Changed

File: main.py (hunk 12 above)
Comment (gcbui) [reply main.py:105]: Removed

File: main.py (hunk 13 above)
Comment (gcbui) [reply main.py:1]: Removed

File: main.py (hunk 4 above)
Comment (gcbui) [reply main.py]: implemented this function

File: main.py (hunk 5 above)
Comment (gcbui) [reply main.py:103]: added function

**Pull request closed without merging**